### ISSUE TITLE
Create an API for integrating with other add-ons

### ISSUE THREAD

**Issue by Mckol** [mbnuqw/sidebery#616]
# Introduction
Since there are plans to create an API for other add-ons, I think it's a good
idea to have a single issue to track its progress.
This will make it much easier for people to find out that this is worked on, and
check how is it going.

## Questions
1. What's the current status of the API? I saw a commit referencing #246 that seems to have added 2 functions if I understand correctly?
2. I have some programming background (Rust, Python) but no real JavaScript experience, would it be possible for me to help implement this?

# Additional information

## Original comment
> I can add API for the other addons, but this will happen after v5 (I've planned big internal changes)
_Originally posted by @mbnuqw in https://github.com/mbnuqw/sidebery/issues/246#issuecomment-633132274_

## Relevant issues
There are multiple use-cases for integrating other extensions with Sidebery, as shown by the following issues:
- #580
- #246
- #279
- #59
- Integration with Tridactyl (no issue)

**Comment by goodboy** [comment]
This is actually one of the main reasons I'm probably going to drop usage for now on a new extension i'm hacking **despite** really preferring the style and startup performance of `sideberry` to others (TST + auto-tab-discard).

[`TST` has really extensive API docs for all sorts of lovely things](https://github.com/piroor/treestyletab/wiki/API-for-other-addons) FWIW that might handy to use as a role model 😉

**Comment by Mikilio** [comment]
I'm really looking forward to integration for tridactyl. I'm ready to help. What is the current roadblock for this?

**Comment by Mikilio** [comment]
My idea is to expose all operations accessible via a `browser.commands` with [runtime.onMessageExternal](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/runtime/onMessageExternal)

**Comment by ChristinWhite** [comment]
This would be such a huge improvement to Sidebery, I've been considering switching to TST for this specific function, but I prefer Sidebery in almost every other way. It would be wonderful to be able to move navigate and move tabs between tab panels by name in a command runner like Surfingkeys.

**Comment by rutgersc** [comment]
@Mikilio Have you looked into this? Would this be something sidebery would be open to accept? I'm also willing to help if so.

**Comment by Mikilio** [comment]
Not really, and it's too far down my priority list to have a realistic chance of that happening, unfortunately.

**Comment by rutgersc** [comment]
@Mikilio I took some time to set up a working example that allows for basic addon interaction, see https://github.com/mbnuqw/sidebery/compare/v5...rutgersc:sidebery:externalMessages

For now I made it so it passes everything through the same API as the one used for keybindings. It looks like all the `browser.commands` are exposed [there](https://github.com/mbnuqw/sidebery/blob/v5/src/services/keybindings.actions.ts#L31) already. The keybinding API currently supports a few more: https://github.com/mbnuqw/sidebery/blob/v5/src/services/keybindings.actions.ts#L158

Example usage in tridactyl

```
command sideberycommand js -d€ browser.runtime.sendMessage("{3c078156-979c-498b-8990-85f7987dd929}", { type: 'command', name: JS_ARGS[1] })€
command sideberycommand.activate sideberycommand activate
command sideberycommand.edit_title sideberycommand edit_title
command sideberycommand.up sideberycommand up
command sideberycommand.up_shift sideberycommand up_shift
command sideberycommand.down sideberycommand down
command sideberycommand.down_shift sideberycommand down_shift
command sideberycommand.group_tabs sideberycommand group_tabs
```
 
Seeing as this change is not too big, would you be willing to accept a PR of this @mbnuqw ?

**Comment by dlight** [comment]
I am writing another extension that needs to get the whole tab structure in real time, as it changes. In TST this is possible with the [API for other addons](https://github.com/piroor/treestyletab/wiki/API-for-other-addons), but in Sidebery it seems this is not possible.

My question is, is this possible only if/when Sidebery provide such an API, or is there some undocumented way to workaround the lack of a public API? Maybe by using CDP rather than writing an extension, I can get Sidebery's tree of tabs?
